src/charm.py

```diff
@@ -5,18 +5,228 @@
 """Charmed operator for the 5G UDM service."""
 
 import logging
+from ipaddress import IPv4Address
+from subprocess import check_output
+from typing import Optional
 
+from charms.observability_libs.v1.kubernetes_service_patch import (  # type: ignore[import]  # noqa: E501
+    KubernetesServicePatch,
+)
+from charms.sdcore_nrf.v0.fiveg_nrf import NRFRequires  # type: ignore[import]
+from jinja2 import Environment, FileSystemLoader
+from lightkube.models.core_v1 import ServicePort
 from ops.charm import CharmBase
+from ops.framework import EventBase
 from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import Layer
 
 logger = logging.getLogger(__name__)
 
+BASE_CONFIG_PATH = "/etc/udm"
+CONFIG_FILE_NAME = "udmcfg.yaml"
+UDM_SBI_PORT = 29503
+NRF_RELATION_NAME = "fiveg_nrf"
+
 
 class UDMOperatorCharm(CharmBase):
     """Charm the service."""
 
     def __init__(self, *args):
         super().__init__(*args)
+        self._container_name = self._service_name = "udm"
+        self._container = self.unit.get_container(self._container_name)
+
+        self.framework.observe(self.on.udm_pebble_ready, self._configure_sdcore_udm)
+
+        self._nrf_requires = NRFRequires(charm=self, relation_name=NRF_RELATION_NAME)
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._configure_sdcore_udm)
+
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[ServicePort(name="sbi", port=UDM_SBI_PORT)],
+        )
+
+    def _configure_sdcore_udm(self, event: EventBase) -> None:
+        """Adds Pebble layer and manages Juju unit status.
+
+        Args:
+            event (EventBase): Juju event.
+        """
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
```

Comment:
Since pebble_ready points here, I don't think we need the defer.

Reply:
Yes, we don't. Removed.